### PR TITLE
add webhook when upgrade cluster 

### PR DESCRIPTION
**Pull request by shuijing198799**
@weekface @xiaojingchen @tennix  PTLA

### CODE REVIEW

File: tests/actions.go [hunk 1]
```diff
@@ -1836,6 +1852,86 @@ func (oa *operatorActions) CheckIncrementalBackup(info *TidbClusterConfig) error
...
+
+func (oa *operatorActions) RegisterWebHookAndService(info *OperatorConfig) error {
+	client := oa.kubeCli
+	glog.Infof("Registering the webhook via the AdmissionRegistration API")
+
+	namespace := "tidb-operator-stability"
+	configName := info.WebhookConfigName
+	filePath := "/webhook.local.config/certificates/ca.crt"
+
+	fd, err := os.Open(filePath)
```
Comment (weekface) [inline tests/actions.go]:
```
defer fd.Close()
```

File: tests/actions.go [hunk 2]
```diff
@@ -1836,6 +1852,86 @@ func (oa *operatorActions) CheckIncrementalBackup(info *TidbClusterConfig) error
...
+
+	// The webhook configuration is honored in 10s.
+	time.Sleep(10 * time.Second)
+
+	return nil
+
+}
+
+func (oa *operatorActions) CleanWebHookAndService(info *OperatorConfig) {
+	oa.kubeCli.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations().Delete(info.WebhookConfigName, nil)
```
Comment (weekface) [inline tests/actions.go]: Don't ignore `error`

File: tests/cmd/stability/main.go [hunk 3]
```diff
@@ -26,8 +28,75 @@ import (
 	"github.com/pingcap/tidb-operator/tests"
 	"github.com/pingcap/tidb-operator/tests/backup"
 	"github.com/pingcap/tidb-operator/tests/pkg/client"
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// toAdmissionResponse is a helper function to create an AdmissionResponse
+// with an embedded error
+func toAdmissionResponse(err error) *v1beta1.AdmissionResponse {
```
Comment (weekface) [inline tests/cmd/stability/main.go]: move all these funcs out of main.go

File: tests/cmd/stability/pods.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/golang/glog"
+	"github.com/pingcap/tidb-operator/tests/pkg/client"
+	"k8s.io/api/admission/v1beta1"
+)
+
+// only allow pods to pull images from specific registry.
```
Comment (weekface) [inline tests/cmd/stability/pods.go]: what is this means?

File: tests/cmd/stability/pods.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		}
+		defer res.Body.Close()
+
+		content, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			glog.Errorf("fail to read response %v", err)
+			return &reviewResponse
+		}
+
+		if strings.Contains(string(content), "\"is_owner\":false") {
```
Comment (weekface) [inline tests/cmd/stability/pods.go]: use `Unmarshal` instead

File: tests/manifests/stability/stability.yaml [hunk 6]
```diff
@@ -23,21 +48,31 @@ kind: Pod
...
   namespace: tidb-operator-stability
   name: tidb-operator-stability
+  labels:
+    app: webhook
 spec:
   serviceAccount: tidb-operator-stability
   containers:
   - name: tidb-operator-stability
-    image: ""
+    image: "hub.pingcap.net/yinliang/pingcap/tidb-operator-stability-test"
```
Comment (weekface) [inline tests/manifests/stability/stability.yaml]: leave this empty

File: tests/actions.go (hunk 1 above)
Comment (shuijing198799) [reply tests/actions.go]: done

File: tests/actions.go [hunk 7]
```diff
@@ -67,7 +69,7 @@ func NewOperatorActions(cli versioned.Interface, kubeCli kubernetes.Interface, c
 
 const (
 	DefaultPollTimeout  time.Duration = 10 * time.Minute
-	DefaultPollInterval time.Duration = 1 * time.Minute
+	DefaultPollInterval time.Duration = 5 * time.Second
```
Comment (weekface) [inline tests/actions.go]: It will be very noisy to use 5 seconds

File: tests/actions.go [hunk 8]
```diff
@@ -1836,6 +1852,91 @@ func (oa *operatorActions) CheckIncrementalBackup(info *TidbClusterConfig) error
 
 }
 
+func strPtr(s string) *string { return &s }
+
+func countEndpoint(e *corev1.Endpoints) int {
```
Comment (weekface) [inline tests/actions.go]: This func is useless.

File: tests/actions.go [hunk 9]
```diff
@@ -1836,6 +1852,91 @@ func (oa *operatorActions) CheckIncrementalBackup(info *TidbClusterConfig) error
...
+	if err := oa.RegisterWebHookAndService(info); err != nil {
+		panic(err)
+	}
+}
+
+func (oa *operatorActions) RegisterWebHookAndService(info *OperatorConfig) error {
+	client := oa.kubeCli
+	glog.Infof("Registering the webhook via the AdmissionRegistration API")
+
+	namespace := "tidb-operator-stability"
```
Comment (weekface) [inline tests/actions.go]: add a `NAMESPACE` env in stability.yaml and use os.Getenv("NAMSPACE") instead.

File: tests/cmd/stability/main.go [hunk 10]
```diff
@@ -43,14 +43,27 @@ func main() {
 	tidbVersion := conf.GetTiDBVersionOrDie()
 	upgardeTiDBVersions := conf.GetUpgradeTidbVersionsOrDie()
 
+	// start a http server in goruntine
+	go func() {
```
Comment (weekface) [inline tests/cmd/stability/main.go]: StartValidatingAdmissionWebhookServerOrDie()

File: tests/cmd/stability/main.go [hunk 11]
```diff
@@ -138,15 +151,15 @@ func main() {
 		oa.DumpAllLogs(operatorCfg, allClusters)
 	}()
 
+	// clean and deploy operator
+	oa.CleanOperatorOrDie(operatorCfg)
+	oa.DeployOperatorOrDie(operatorCfg)
+
```
Comment (weekface) [inline tests/cmd/stability/main.go]: why move them here?

File: tests/cmd/stability/main.go [hunk 12]
```diff
@@ -185,6 +198,9 @@ func main() {
 	oa.CheckTidbClusterStatusOrDie(cluster1)
 	oa.CheckTidbClusterStatusOrDie(cluster2)
 
+	// before upgrade cluster, deploy and register webhook first
```
Comment (weekface) [inline tests/cmd/stability/main.go]:
```suggestion
	// before upgrade cluster, register webhook first
```

File: tests/cmd/stability/pods.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+	IsOwner bool `json:"is_owner"`
+}
+
+// only allow pods to be delete when it is not ddlowner of tidb, not leader of pd and not
+// master of tikv.
+func admitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	glog.Infof("admitting pods")
+	httpClient := &http.Client{}
+	podResource := metav1.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}
+	if ar.Request.Resource != podResource {
```
Comment (weekface) [inline tests/cmd/stability/pods.go]: You have register the webhook with [pods deletion](https://github.com/pingcap/tidb-operator/pull/389/files#diff-67dacabf080ca4dc8e95d2acf12ee36bR1905). I think there is no need to verify it again,

File: tests/cmd/stability/main.go (hunk 11 above)
Comment (shuijing198799) [reply tests/cmd/stability/main.go]: the webhook delete is follow operator, if the test is terminate the webhook will not be clean. next run can't delete pod.

File: tests/cmd/stability/main.go (hunk 10 above)
Comment (shuijing198799) [reply tests/cmd/stability/main.go]: config is just associate with wehook and k8s action , so register is better.

File: tests/cmd/stability/main.go (hunk 11 above)
Comment (weekface) [reply tests/cmd/stability/main.go]: Got it.

File: tests/cmd/stability/route.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// admitFunc is the type we use for all of our validators and mutators
+type admitFunc func(v1beta1.AdmissionReview) *v1beta1.AdmissionResponse
+
+// serve handles the http portion of a request prior to handing to an admit
+// function
+func serve(w http.ResponseWriter, r *http.Request, admit admitFunc) {
+	var body []byte
+	if r.Body != nil {
```
Comment (weekface) [inline tests/cmd/stability/route.go]: What about `r.Body` is nil?

File: tests/manifests/stability/stability.yaml [hunk 15]
```diff
@@ -32,17 +59,27 @@ spec:
...
+    - name: webhook-certs
+      readOnly: true
+      mountPath: /webhook.local.config/certificates
     env:
     - name: MY_NODE_NAME
       valueFrom:
         fieldRef:
           fieldPath: spec.nodeName
+    - name: NAMESPACE
+	  value: tidb-operator-stability
```
Comment (weekface) [inline tests/manifests/stability/stability.yaml]:
```
      valueFrom:
        fieldRef:
          fieldPath: metadata.namespace
`

File: tests/cmd/stability/main.go [hunk 16]
```diff
@@ -45,14 +45,20 @@ func main() {
 	tidbVersion := conf.GetTiDBVersionOrDie()
 	upgardeTiDBVersions := conf.GetUpgradeTidbVersionsOrDie()
 
+	// start a http server in goruntine
+	go oa.StartValidatingAdmissionWebhookServerOrDie()
```
Comment (xiaojingchen) [inline tests/cmd/stability/main.go]: this could start before upgrade case running, not need to start too early.

**Pull request merged**